### DISCUSSION TITLE
shadowhook v1.1.1

### DISCUSSION THREAD

**Discussion by Colibrow** [bytedance/android-inline-hook#82] [Announcements]
# New Features

### 1. Added 4 APIs for registering/unregistering `pre/post` callbacks of `soinfo::call_constructors` and `soinfo::call_destructors`

- `shadowhook_register_dl_init_callback()`, `shadowhook_unregister_dl_init_callback()`.
- `shadowhook_register_dl_fini_callback()`, `shadowhook_unregister_dl_fini_callback()`.

# Bug Fixes

### 1. Fixed the IT instruction fix bug under Thumb instructions
The fix for the IT instruction under the `ne` condition under Thumb has been corrected.

In the previous version, because the incorrect use of memset would make `it->firstcond` always 0, only the `eq` case could be fixed correctly in the conditional judgment.

* Related issue: #80
* Thanks: @i0rn

### 2. Fix the memory leak bug caused by the thread re-entering the proxy function
Currently, if the function after `pthread_key_clean_all` is hooked, the proxy function execution will be skipped and no memory leak will be caused.

In the previous version: if the proxy function is run after the thread's `pthread_key_clean_all` call, if there are more than `SH_HUB_THREAD_MAX` cached `hub_stack` at this time, new memory will be mmaped out through `sh_hub_stack_create` and will not be released.
# Improvement

- Upgrade the versions of NDK, CMake, gradle and AGP.

# 新特性

### 1. 新增4个API，用于注册 / 反注册 `soinfo::call_constructors` 和`soinfo::call_destructors` 的 `pre /post` 回调

- `shadowhook_register_dl_init_callback()`，`shadowhook_unregister_dl_init_callback()`。
- `shadowhook_register_dl_fini_callback()`，`shadowhook_unregister_dl_fini_callback()`。

# Bugs 修复

### 1. 修复了Thumb指令下的IT指令修复bug
针对IT指令在Thumb下的`ne`条件下的指令修复进行了修正。

在之前的版本中因为memset的错误使用会使得`it->firstcond`一直为0，所以在条件判断中只能正确修复`eq`的情况。

* 相关的 issue：#80
* 感谢：@i0rn

### 2. 修复线程再次进入代理函数引起的内存泄漏bug
目前如果hook了在`pthread_key_clean_all`之后的函数将会跳过代理函数执行，不引起内存泄漏。

在之前的版本中：如果代理函数在线程的`pthread_key_clean_all`调用之后运行，如果此时已经超过`SH_HUB_THREAD_MAX`个缓存的`hub_stack`，会通过`sh_hub_stack_create` mmap出新的内存并且不会释放。
# 改进

- 升级 NDK，CMake，gradle 和 AGP 的版本。

<hr /><em>This discussion was created from the release <a href='https://github.com/bytedance/android-inline-hook/releases/tag/v1.1.1'>shadowhook v1.1.1</a>.</em>

**Comment by GeekMS** [comment]
Hello,

I have been reading the ShadowHook source code and noticed the `libshadowhook_nothing.so` library that was introduced in v1.1.1.

I couldn't find any documentation, comments, or release notes explaining its purpose. The only related information I found was issue #91, which mentions that this library was added and should be included when building from source.

From reading `sh_linker.c`, it appears that ShadowHook hooks `soinfo::call_constructors()` / `call_destructors()`, then calls:

```cpp
dlopen("libshadowhook_nothing.so", RTLD_NOW);
```

and later scans the corresponding `soinfo` object to determine field offsets.

My questions are:

1. What is the exact purpose of `libshadowhook_nothing.so`?
2. Why is it necessary to load a separate shared library instead of using `libshadowhook.so` itself?
3. Is its only purpose to trigger creation of a fresh `soinfo` instance for offset discovery, or is there another reason behind this design?
4. Does the library itself need to contain anything special, or is an almost-empty shared library sufficient?

I would appreciate any explanation about the design rationale, since I couldn't find any documentation describing this mechanism.

Thank you.

**Reply by HanSoBored** [reply]
as far i know, its like "trigger" for scan private memory from linker android, it does nothing  libshadowhook.so already load and need another shared library load to find offset field, in sh_nothing.c it just static int a;

**Reply by maxxu68** [reply]
It's a bootstrap .so. shadowhook dlopen()s a not-yet-loaded library at init to get a fresh soinfo * and scan out the field offsets it needs — libshadowhook.so itself won't work since it's already loaded. Empty is fine (static int a;); it's what makes the dl_init/dl_fini callbacks work, which is what bytehook uses.
